### ISSUE TITLE
Experimental results reproduction

### ISSUE THREAD

**Issue by IbsenChan** [facebookresearch/deepcluster#32]
Hello,

I come across your paper and find it simple & useful. I try to reproduce the experimental results in the paper by your released code without any changes, but I find that NMI converges from 0.54 to 0.68 very soon after dozens of epochs. The NMI training curve is not that like Fig. 2 (b) in the paper so as the linear classification experiment in Table 1. I know the core of your algorithm is not only the pseudo labels provided by k-means but also the data augmentation/transformation invariance charateristic from your training settings ("We cluster the central cropped images features and perform data augmentation (random horizontal flips and crops of random sizes and aspect ratios) when training the network"). Did I miss some important settings? Can you help me point out the problem? It will help me a lot in my research and I hope to cite your paper.

Thanks~

**Comment by mathildecaron31** [comment]
Hi, thanks for your interest.
The code as it is should reproduce the results of the paper.
Indeed data augmentation is crucial for the network to learn good transferable features.
You should set sobel flag to 1 and.
Moreover, the classification results of Table 1 are obtained after 500 epochs of deepcluster.

**Comment by IbsenChan** [comment]
Thank you very much for your prompt reply!

I use the default settings in the code, 256 batch size, 0.05 constant learning rate. But the NMI converges to 0.70 very quickly. In Fig.2 (b) in the paper, NMI approached to 0.70 only after 150epochs. I am not sure if the feature maps will get more discriminated after 500 epochs if the NMI always maintains in a same level.

![24epoch](https://user-images.githubusercontent.com/7579019/58927865-e0e34600-8782-11e9-9a68-067301bb89ad.PNG)
![45epoch](https://user-images.githubusercontent.com/7579019/58933154-efd4f300-8798-11e9-9b78-cf4d5f234e49.PNG)

Actually, before this experiment, I try a large batch size 1536 with 0.3 learning rate accordingly for fast training in 6 GPUs. The NMI always maintains in a range of [0.68, 0.70] from 10th epoch to 210th epochs, and the feature seem not get more discriminated. I guess the possible reason may be that the NMI converges too fast which hinders the network to learn more discriminated representation. 

What is the main operation to stop NMI converge too fast and let the network learn more?

**Comment by mathildecaron31** [comment]
can you send me all the used arguments?
do you use sobel filtering (`--sobel`) in the bash command ?

**Comment by IbsenChan** [comment]
I used the default settings as follows:

![args](https://user-images.githubusercontent.com/7579019/58953504-465a2580-87c9-11e9-9b37-55e7651afb24.PNG)

**Comment by WangFeng18** [comment]
I have encountered the same problem, have you figured out why the value of NMI converge so fast.@IbsenChan. I ran the code with default setting and the NMI value reached around 0.68 after 10 epochs.

**Comment by IbsenChan** [comment]
Nope. @WangFeng18

**Comment by WangFeng18** [comment]
@IbsenChan Thanks. What about your evaluation results, such as accuracy on ImageNet and results of fine-tuning on VOC2007 dataset.

**Comment by IbsenChan** [comment]
@WangFeng18 Run eval_linear.sh on ImageNet with --conv 5 and only get 22% accuracy (the pretrained model used here wasn't obtained after 500 epochs of deepcluster). What about you?

**Comment by WangFeng18** [comment]
![image](https://user-images.githubusercontent.com/43294876/59399787-e3b7e980-8dc7-11e9-9603-a9e3fe9ab21d.png)
@IbsenChan This is my result on running the eval_linear.sh, I only got about 22% Top 1 accuracy on Training Set, and only 3.5% on Validation Set. And the finetuning results on VOC2007 is about 53%, my model was trained after 100 epochs.

**Comment by sbrugman** [comment]
@IbsenChan @WangFeng18 The same results over here, NMI around 0.7 after 15 epochs:
```
###### Epoch [15] ######
Time: 7351.146 s
Clustering loss: 797736.688
ConvNet loss: 4.850
NMI against previous assignment: 0.711
#######################
```

**Comment by mathildecaron31** [comment]
Hi, thank you so much for raising this issue. I have figured out the difference between this code version and the version I used for my DeepCluster experiments.

Unlike the version I used for my final experiments, this version has a *constant faiss seed*.

As a consequence, at every clustering the same features are picked as initialization centroids. As k-means is quite sensible to initialization and the features don't change a lot from an epoch to another, the clusterings tend to be very similar from an epoch to another.

By changing the faiss seed (thus changing the picked initialization centroids) at every epoch, I reproduce the results of my paper.

I am going to commit this fix promptly.
